cdap-unit-test/src/test/java/co/cask/cdap/test/app/WorkflowAppWithLocalDatasets.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.test.app;
+
+import co.cask.cdap.api.annotation.UseDataSet;
+import co.cask.cdap.api.app.AbstractApplication;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.api.dataset.lib.KeyValueTable;
+import co.cask.cdap.api.mapreduce.AbstractMapReduce;
+import co.cask.cdap.api.mapreduce.MapReduceContext;
+import co.cask.cdap.api.workflow.AbstractWorkflow;
+import co.cask.cdap.api.workflow.AbstractWorkflowAction;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.IntWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+import java.util.StringTokenizer;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Workflow application containing local datasets.
+ */
+public class WorkflowAppWithLocalDatasets extends AbstractApplication {
+  public static final String WORDCOUNT_DATASET = "wordcount";
+  public static final String RESULT_DATASET = "result";
+  @Override
+  public void configure() {
+    setName("WorkflowAppWithLocalDatasets");
+    setDescription("App to test the local dataset functionality for the Workflow.");
+
+    addMapReduce(new WordCount());
+    addWorkflow(new WorkflowWithLocalDatasets());
+    createDataset(RESULT_DATASET, KeyValueTable.class);
+  }
+
+  /**
+   * Workflow which configures the local dataset.
+   */
+  public static class WorkflowWithLocalDatasets extends AbstractWorkflow {
+
+    @Override
+    protected void configure() {
+      setName("WorkflowWithLocalDatasets");
+      setDescription("Workflow program with local datasets.");
+      createLocalDataset(WORDCOUNT_DATASET, KeyValueTable.class);
+      addMapReduce("WordCount");
+      addAction(new LocalDatasetReader());
```

Comment:
Wanted to get feedback on the MapReduce programs first. As mentioned in the PR description will add the Spark related changes in the next PR.